src/linear_algebra/char_poly/coeff.lean

```diff
@@ -0,0 +1,191 @@
+  intros, apply nat_degree_char_matrix_val_le,
+end
+
+lemma high_coeff_char_poly_eq_coeff_prod_diag {k : ℕ} (h : fintype.card n - 1 ≤ k) :
+  (char_poly M).coeff k = (univ.prod (λ (i : n), X - C (M i i))).coeff k :=
+begin
+  apply eq_of_sub_eq_zero, rw ← coeff_sub, apply polynomial.coeff_eq_zero_of_degree_lt,
+  apply lt_of_lt_of_le (low_degree_char_poly_sub_diagonal M) _, rw with_bot.coe_le_coe, apply h,
+end
+
+lemma det_of_dim_zero (h : fintype.card n = 0) (M : matrix n n R) : M.det = 1 :=
+by { rw fintype.card_eq_zero_iff at h, suffices : M = 1, { simp [this] }, ext, tauto }
+
+
+theorem degree_char_poly_eq_dim [nontrivial R] (M: matrix n n R) :
+(char_poly M).degree = fintype.card n :=
+begin
+  by_cases fintype.card n = 0, rw h, unfold char_poly, rw det_of_dim_zero, simpa,
+  rw ← sub_add_cancel (char_poly M) (∏ (i : n), (X - C (M i i))),
+  have h1 : (∏ (i : n), (X - C (M i i))).degree = fintype.card n,
+  { rw degree_eq_iff_nat_degree_eq_of_pos, swap, apply nat.pos_of_ne_zero h,
+    rw nat_degree_prod', simp_rw nat_degree_X_sub_C, unfold fintype.card, simp,
+    simp_rw (monic_X_sub_C _).leading_coeff, simp, },
+  rw degree_add_eq_of_degree_lt, exact h1, rw h1,
+  apply lt_trans (low_degree_char_poly_sub_diagonal M), rw with_bot.coe_lt_coe,
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+theorem nat_degree_char_poly_eq_dim [nontrivial R] (M: matrix n n R) :
+  (char_poly M).nat_degree = fintype.card n :=
+nat_degree_eq_of_degree_eq_some (degree_char_poly_eq_dim M)
+
+lemma char_poly_monic_of_nontrivial [nontrivial R] (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  by_cases fintype.card n = 0, rw [char_poly, det_of_dim_zero h], apply monic_one,
+  have mon : (∏ (i : n), (X - C (M i i))).monic,
+  { apply monic_prod_of_monic univ (λ i : n, (X - C (M i i))), simp [monic_X_sub_C], },
+  rw ← sub_add_cancel (∏ (i : n), (X - C (M i i))) (char_poly M) at mon,
+  rw monic at *, rw leading_coeff_add_of_degree_lt at mon, rw ← mon,
+  rw degree_char_poly_eq_dim, rw ← neg_sub, rw degree_neg,
+  apply lt_trans (low_degree_char_poly_sub_diagonal M), rw with_bot.coe_lt_coe,
+  rw ← nat.pred_eq_sub_one, apply nat.pred_lt, apply h,
+end
+
+lemma char_poly_monic (M : matrix n n R) :
+  monic (char_poly M) :=
+begin
+  classical, by_cases h : nontrivial R,
+  { letI := h, apply char_poly_monic_of_nontrivial, },
+  { rw nontrivial_iff at h, push_neg at h, apply h, }
+end
+
+--shouldn't need these instances, but might need casework
+theorem trace_from_char_poly [nontrivial R] [nonempty n] (M: matrix n n R) :
+(matrix.trace n R R) M = -(char_poly M).coeff (fintype.card n - 1) :=
+begin
+  rw high_coeff_char_poly_eq_coeff_prod_diag, swap, refl,
+  rw [fintype.card, prod_X_sub_C_coeff_card_pred univ (λ i : n, M i i)], simp,
+  rw [← fintype.card, fintype.card_pos_iff], apply_instance,
+end
+
+lemma ring_hom_det {S : Type w} [comm_ring S] {M : matrix n n R} {f : R →+* S} :
+  f M.det = matrix.det (f.map_matrix M) :=
```

Comment:
Perhaps `ring_hom.map_det`, so it can be used with dot notation? (Maybe requires making `f` explicit.)

Reply:
This should move back to `data.matrix.basic`.